system/src/system_task.cpp

```diff
@@ -368,12 +368,13 @@ void system_delay_ms_non_threaded(unsigned long ms, bool force_no_background_loo
 
 void system_delay_ms(unsigned long ms, bool force_no_background_loop=false)
 {
-    if (system_thread_get_state(NULL)!=0) {
-        HAL_Delay_Milliseconds(ms);
+    if (system_thread_get_state(NULL) == spark::feature::DISABLED &&
```

Comment:
Yes, the ApplicationThread / SystemThread active objects don't have a meaning when threading is disabled (yet the objects still exist), but SYSTEM_THREAD_CURRENT / APPLICATION_THREAD_CURRENT have a clear meaning.

To explain, I'll work back from the issue I had.

`delay()` should only call `Spark_Idle_Events` when called from the user's `setup()/loop()` and threading is disabled. It should not be called from the system thread or from a wiring thread.

How to make the difference for that last case? Check if the current thread is the application thread.

When threading is disabled, the application thread and system thread are the original one created by WICED. They just happen not to be running a deferred execution loop.

So in main.cpp, the ApplicationThread and SystemThread are associated with the current thread if threading is disabled so that SYSTEM_THREAD_CURRENT / APPLICATION_THREAD_CURRENT return the correct result with Wiring threads.

Similarly if PLATFORM_THREADING is not compiled in, SYSTEM_THREAD_CURRENT / APPLICATION_THREAD_CURRENT always return true.

With this change, you can drop the check of `isStarted` in the `SYSTEM_THREAD_CONTEXT_...` macros because `isCurrentThread` will return the correct result whether threading is enabled or not. Do you agree?